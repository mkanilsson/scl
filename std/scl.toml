[package]
name = "std"
description = "Scl's standard library"
type = "lib"
