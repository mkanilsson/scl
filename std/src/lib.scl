extern proc exit(i32) !
