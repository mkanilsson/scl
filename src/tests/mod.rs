use std::{env, fs, path::PathBuf, process::Command};

use insta::glob;
use miette::{GraphicalReportHandler, GraphicalTheme};
use nanoid::nanoid;

use crate::{codegen::Codegen, package::Package, typechecker::Checker};

#[test]
fn runtime() {
    let std_package = Package::from_path("std", "std".into()).unwrap();
    let std_package = std_package.parse().unwrap();

    fn generate_tempfile(extension: &str) -> PathBuf {
        let name = format!("{}.{extension}", nanoid!());
        env::temp_dir().join(name)
    }

    glob!("sources/runtime/*", |file| {
        let main_package = Package::from_file(file.into()).unwrap();
        let main_package = main_package.parse().unwrap();

        let mut checker = Checker::new();
        let checked_std_package = checker.add_package(&std_package, &vec![]).unwrap();

        let checked_main_package = checker
            .add_package(&main_package, &vec![(
                "std".into(),
                checked_std_package.package_id,
            )])
            .unwrap();

        insta::assert_debug_snapshot!(checked_main_package);

        let all_units = checked_std_package
            .units
            .into_iter()
            .chain(checked_main_package.units.into_iter())
            .collect::<Vec<_>>();

        let mut codegener = Codegen::new(all_units, checker);
        let code = codegener.generate();

        insta::assert_snapshot!(code);

        let tmp = generate_tempfile("ssa");
        let asm = generate_tempfile("S");
        let executable = generate_tempfile("out");

        std::fs::write(&tmp, code).unwrap();

        let qbe_cmd = Command::new("qbe")
            .arg("-o")
            .arg(&asm)
            .arg(tmp)
            .output()
            .unwrap();

        if !qbe_cmd.status.success() {
            let stdout = String::from_utf8_lossy(&qbe_cmd.stdout);
            let stderr = String::from_utf8_lossy(&qbe_cmd.stderr);

            panic!("QBE failed:\nOut:\n{stdout}Err:\n{stderr}");
        }

        let gcc_cmd = Command::new("gcc")
            .arg("-o")
            .arg(&executable)
            .arg(&asm)
            .output()
            .unwrap();

        if !gcc_cmd.status.success() {
            let stdout = String::from_utf8_lossy(&gcc_cmd.stdout);
            let stderr = String::from_utf8_lossy(&gcc_cmd.stderr);

            panic!("gcc failed:\nOut:\n{stdout}Err:\n{stderr}");
        }

        let run_cmd = Command::new(executable).output().unwrap();

        let stdout = String::from_utf8_lossy(&run_cmd.stdout);
        let stderr = String::from_utf8_lossy(&run_cmd.stderr);

        insta::assert_snapshot!(stdout);
        insta::assert_snapshot!(stderr);
        insta::assert_snapshot!(run_cmd.status);
    });
}

#[test]
fn typecheck_errors() {
    let std_package = Package::from_path("std", "std".into()).unwrap();
    let std_package = std_package.parse().unwrap();

    glob!("sources/typecheck_error/*", |file| {
        let main_package = Package::from_file(file.into()).unwrap();
        let main_package = main_package.parse().unwrap();

        let mut checker = Checker::new();
        let checked_std_package = checker.add_package(&std_package, &vec![]).unwrap();

        let err = checker
            .add_package(&main_package, &vec![(
                "std".into(),
                checked_std_package.package_id,
            )])
            .err()
            .unwrap();

        let mut buf = String::new();
        GraphicalReportHandler::new_themed(GraphicalTheme::none())
            .render_report(&mut buf, miette::Report::new(err).as_ref())
            .unwrap();

        insta::assert_snapshot!(buf);
    });
}

#[test]
fn parse_expr() {
    glob!("sources/parse_expr/*", |file| {
        let source = fs::read_to_string(file).unwrap();
        let expr = crate::new_parser::Parser::parse_expr(&source);

        insta::assert_debug_snapshot!(expr);
    });
}

#[test]
fn parse_stmt() {
    glob!("sources/parse_stmt/*", |file| {
        let source = fs::read_to_string(file).unwrap();
        let expr = crate::new_parser::Parser::parse_stmt(&source);

        insta::assert_debug_snapshot!(expr);
    });
}
