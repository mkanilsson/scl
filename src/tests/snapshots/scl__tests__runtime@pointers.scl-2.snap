---
source: src/tests/mod.rs
expression: code
input_file: src/tests/sources/runtime/pointers.scl
---
type :Thing = { l }
type :Other = { w }
export function w $add(w %a, w %b) {
@start
	%.ss.0 =l alloc4 4
	%.ss.1 =l alloc4 4
	storew %a, %.ss.0
	storew %b, %.ss.1
@body
	%sv.1 =w loadw %.ss.0
	%sv.2 =w loadw %.ss.1
	%binop.3 =w add %sv.1, %sv.2
	ret %binop.3
}
export function w $main() {
@start
	%.ss.0 =l alloc4 4
	%.ss.1 =l alloc8 8
	%.ss.2 =l alloc8 8
	%.ss.3 =l alloc8 8
	%.ss.4 =l alloc4 4
	%.ss.5 =l alloc4 4
	%.ss.6 =l alloc8 8
	%.ss.7 =l alloc8 8
	%.ss.8 =l alloc4 4
	%.ss.9 =l alloc4 4
	%.ss.10 =l alloc8 8
	%.ss.11 =l alloc8 8
	%.ss.12 =l alloc8 8
	%.ss.13 =l alloc4 4
	%.ss.14 =l alloc4 4
	%.ss.15 =l alloc4 4
	%.ss.16 =l alloc4 4
	%.ss.17 =l alloc8 8
	%.ss.18 =l alloc8 8
	%.ss.19 =l alloc4 4
	%.ss.20 =l alloc4 4
@body
	storew 10, %.ss.0
	# Store value into Thing.value (offset: 0)
	%offset.4 =l add %.ss.1, 0
	storel %.ss.0, %offset.4
	blit %.ss.1, %.ss.2, 8
	%.member_access.value.7 =l add %.ss.2, 0
	%.deref.ptr.8 =l loadl %.member_access.value.7
	%.ss.3 =l loadl %.deref.ptr.8
	%printf.return_value.5 =w call $printf(l $str.6, ..., l %.ss.3)
	storew 20, %.ss.5
	storel %.ss.5, %.ss.6
	%.deref.ptr.11 =l loadl %.ss.6
	%.ss.7 =l loadl %.deref.ptr.11
	%printf.return_value.9 =w call $printf(l $str.10, ..., l %.ss.7)
	storew 30, %.ss.9
	# Store value into Thing.value (offset: 0)
	%offset.12 =l add %.ss.10, 0
	storel %.ss.9, %offset.12
	%offset.13 =l add %.ss.10, 0
	%member_access.14 =l loadl %offset.13
	storel %member_access.14, %.ss.12
	%.deref.ptr.15 =l loadl %.ss.12
	%.ss.11 =l loadl %.deref.ptr.15
	storel %.ss.11, %.ss.13
	%sv.18 =w loadw %.ss.13
	%printf.return_value.16 =w call $printf(l $str.17, ..., w %sv.18)
	# Store value into Other.value (offset: 0)
	%offset.19 =l add %.ss.15, 0
	storew 40, %offset.19
	blit %.ss.15, %.ss.16, 4
	storel %.ss.16, %.ss.17
	%.deref.ptr.20 =l loadl %.ss.17
	%.ss.18 =l loadl %.deref.ptr.20
	storel %.ss.18, %.ss.19
	%printf.return_value.21 =w call $printf(l $str.22, ..., :Other %.ss.19)
	ret 0
}
data $str.6 = { b "a.value.* = %d\n", b 0 }
data $str.10 = { b "c.* = %d\n", b 0 }
data $str.17 = { b "e = %d\n", b 0 }
data $str.22 = { b "h = %d\n", b 0 }
